master/internal/stream/system.go

```diff
@@ -218,37 +277,79 @@ func (ps PublisherSet) Websocket(socket *websocket.Conn, c echo.Context) error {
 
 		// any modifications to our subscriptions?
 		for _, mod := range mods {
-			temp, err := ss.SubscriptionMod(mod, ctx)
+			temp, err := ss.SubscriptionMod(ctx, mod)
 			if err != nil {
 				return errors.Wrapf(err, "error modifying subscriptions")
 			}
 			msgs = append(msgs, temp...)
-			// TODO: also append a sync message (or one sync per SubscriptionModMsg)
+			// XXX: also append a sync message (or one sync per SubscriptionModMsg)
 		}
 
 		// write msgs to the websocket
 		err = writeAll(socket, msgs)
 		if err != nil {
-			// TODO: don't log broken pipe errors.
+			// XXX: don't log broken pipe errors.
 			if err != nil {
 				return errors.Wrapf(err, "error writing to socket")
 			}
 		}
 	}
+}
 
-	return nil
+func (ps *PublisherSet) permissionChangeLoop(ctx context.Context) {
+	permListener, err := AuthZProvider.Get().GetPermissionChangeListener()
+	switch {
+	case err != nil:
+		log.Errorf("error occurred while getting permission change listener: %s", err)
+		fallthrough
+	case permListener == nil:
+		// no need to loop or cleanup if there's no listener
+		return
+	}
+	defer func() {
+		err := permListener.Close()
+		if err != nil {
+			log.Debugf("error occurred while closing permission listener: %s", err)
+		}
+	}()
+
+	for {
+		select {
+		// did permissions change?
+		case <-permListener.Notify:
+			log.Debugf("permission change detected, restarting publisher set")
+			for _, err := range ps.Restart() {
+				if err != nil {
+					log.Debugf("error restarting publisher set: %v", err)
+				}
+			}
+		// is the listener still alive?
+		case <-time.After(30 * time.Second):
+			pingErrChan := make(chan error)
+			go func() {
+				err = permListener.Ping()
+				pingErrChan <- errors.Wrap(err, "no active connection")
+			}()
+			if err := <-pingErrChan; err != nil {
+				log.Errorf("permission listener failed: %s", err)
```

Comment:
This also feels off, should there be a recovery/shutdown mechanism in the event that there is not active connection on the permission listener?

Reply:
makes sense, this can be addressed by handling errors that come out of `permissionChangeLoop()` and having this return an error.